server/src/main/java/org/elasticsearch/index/mapper/SearchFields.java

```diff
@@ -0,0 +1,153 @@
+        this.mapperService = mapperService;
+        this.allowUnmappedFields = allowUnmappedFields;
+    }
+
+    public void setAllowUnmappedFields(boolean allowUnmappedFields) {
+        this.allowUnmappedFields = allowUnmappedFields;
+    }
+
+    public void setMapUnmappedFieldAsString(boolean mapUnmappedFieldAsString) {
+        this.mapUnmappedFieldAsString = mapUnmappedFieldAsString;
+    }
+
+    public ParsedDocument parseDocument(SourceToParse source) throws MapperParsingException {
+        return mapperService.documentMapper() == null ? null : mapperService.documentMapper().parse(source);
+    }
+
+    public FieldNameAnalyzer getFieldNameIndexAnalyzer() {
+        DocumentMapper documentMapper = mapperService.documentMapper();
+        return documentMapper == null ? null : documentMapper.mappers().indexAnalyzer();
+    }
+
+    public boolean hasNested() {
+        return mapperService.hasNested();
+    }
+
+    public boolean hasMappings() {
+        return mapperService.documentMapper() != null;
+    }
+
+    /**
+     * Returns all the fields that match a given pattern. If prefixed with a
+     * type then the fields will be returned with a type prefix.
+     */
+    public Set<String> simpleMatchToIndexNames(String pattern) {
+        return mapperService.simpleMatchToFullName(pattern);
+    }
+
+    /**
+     * Returns the {@link MappedFieldType} for the provided field name.
+     * If the field is not mapped, the behaviour depends on the index.query.parse.allow_unmapped_fields setting, which defaults to true.
+     * In case unmapped fields are allowed, null is returned when the field is not mapped.
+     * In case unmapped fields are not allowed, either an exception is thrown or the field is automatically mapped as a text field.
+     *
+     * @throws QueryShardException if unmapped fields are not allowed and automatically mapping unmapped fields as text is disabled.
+     * @see QueryShardContext#setAllowUnmappedFields(boolean)
+     * @see QueryShardContext#setMapUnmappedFieldAsString(boolean)
+     */
+    public MappedFieldType fieldType(String name) {
+        return failIfFieldMappingNotFound(name, mapperService.fieldType(name));
+    }
+
+    /**
+     * Returns true if the field identified by the provided name is mapped, false otherwise
+     */
+    public boolean isFieldMapped(String name) {
+        return mapperService.fieldType(name) != null;
+    }
+
+    //TODO this is scary here as it allows to look up mappers which are not aware of runtime fields
+    public ObjectMapper getObjectMapper(String name) {
+        return mapperService.getObjectMapper(name);
+    }
+
+    /**
+     * Given a type (eg. long, string, ...), returns an anonymous field type that can be used for search operations.
+     * Generally used to handle unmapped fields in the context of sorting.
+     */
+    public MappedFieldType buildAnonymousFieldType(String type) {
```

Comment:
I initially wasn't planning on including all these methods. Then I realized that we can make the query shard context not depend on MapperService which felt like a good idea. I am now wondering though if this isn't too heavy to carry around. Maybe we want to use different abstractions, smaller, for aggs and fetch.

Reply:
one idea could be, as a follow-up to introduce an interface only for the field lookups (fieldType and isMappedField) and see if that is enough to be carried around in some places rather than the whole SearchFields. Also, I want to see if we can reduce the usages of getObjectMapper and possibly expose something specific to nested instead given that it's its biggest user.

Reply:
I think abstracting all of the MapperService access into this one class makes it too heavy.  Let's try and keep it focussed on just mediating access to MappedFieldType? So `simpleMatchToIndexNames`, `fieldType` and `isFieldMapped` only.  I think I'd also prefer it as an interface that `QueryShardContext` and `AggregationContext` implement, which should make the change much less invasive.

What this does highlight is that QueryShardContext itself is something of a monster to be passing around, and that it might be worth trying to factor out some smaller interfaces.  `SearchFields` is a start, but for example query rewriting and query building probably don't need access to the whole QSC object.  Nested object access can probably be reworked here as well, and some of the methods that use MapperService here can probably be removed - `buildAnonymousFieldType` feels to me like it's just the wrong abstraction to use for sorting on unmapped fields, and the `allowUnmappedFields` and `setMapUnmappedFieldsAsText` methods are only used by the percolator, which could just wrap QueryShardContext and override the relevant methods instead.

Reply:
I prefer this as its own class rather than an interface, I think. I haven't read the entire change yet, but in the past few years I've been leaning more towards encapsulation over inheritance in general. I'll keep reading some more but I wanted to throw that out there because I bumped into this comment.